### PR TITLE
[TTTE-14] Get practice history by test id and user id API & [TTTE-46] Submit answer by question id API

### PR DESCRIPTION
**Pull request by vtronghieu**

### CODE REVIEW

**Review by NguyenDinhBuiDuy — CHANGES_REQUESTED** [review]

File: src/main/java/com/tip/dg4/toeic_exam/dto/PartTestDto.java [hunk 1]
```diff
@@ -17,4 +17,6 @@ public class PartTestDto {
     private UUID practicePartId;
     private String type;
     private String name;
+    private Integer correctAnswer;
+    private int totalQuestion;
```
[the remark this answers is not in the document]
Comment (minhtrungs013) [reply src/main/java/com/tip/dg4/toeic_exam/dto/PartTestDto.java]: I fix it

File: src/main/java/com/tip/dg4/toeic_exam/models/TestHistoryStatus.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    private String value;
+
+    public static TestHistoryStatus getStatus(String status) {
+        for (TestHistoryStatus testHistoryStatus : TestHistoryStatus.values()) {
+            if (status.equalsIgnoreCase(testHistoryStatus.getValue())) {
+                return testHistoryStatus;
+            }
+        }
+        return null;
```
[the remark this answers is not in the document]
Comment (minhtrungs013) [reply src/main/java/com/tip/dg4/toeic_exam/models/TestHistoryStatus.java:22]: I fix it

File: src/main/java/com/tip/dg4/toeic_exam/services/implement/PartTestServiceImpl.java [hunk 3]
```diff
@@ -52,13 +67,27 @@ public void createPartTest(PartTestDto partTestDto) {
...
         }
         List<PartTest> partTests = partTestRepository.findByPracticePartId(practicePartId);
+        List<PartTestDto> testDtoList = partTests.stream().map(partTestMapper::convertModelToDto).collect(Collectors.toCollection(CopyOnWriteArrayList::new));
 
-        return partTests.stream().map(partTestMapper::convertModelToDto).toList();
+        testDtoList.parallelStream().forEachOrdered(testDto -> {
+            List<TestHistory> testHistories = testHistoryService.getTestHistoriesByTestIdAndUserId(userId, testDto.getId());
+
+            if (!testHistories.isEmpty()) {
+                int lastIndex = testHistories.size() - 1;
```
[the remark this answers is not in the document]
Comment (minhtrungs013) [reply src/main/java/com/tip/dg4/toeic_exam/services/implement/PartTestServiceImpl.java:77]: I fix it

File: src/main/java/com/tip/dg4/toeic_exam/mappers/UserAnswerMapper.java [hunk 4]
```diff
@@ -19,23 +19,23 @@ public UserAnswer convertDtoToModel(UserAnswerDto userAnswerDto) {
...
     public UserAnswerDto convertModelToDto(UserAnswer userAnswer) {
         UserAnswerDto userAnswerDto = new UserAnswerDto();
-        ListenAnswerDto listenAnswerDto = listenAnswerMapper.convertModelToDto(userAnswer.getListenAnswer());
-
-        userAnswerDto.setQuestionId(userAnswer.getQuestionId());
-        userAnswerDto.setOptionAnswer(userAnswer.getOptionAnswer());
-        userAnswerDto.setListenAnswer(listenAnswerDto);
-        userAnswerDto.setWriteAnswer(userAnswer.getWriteAnswer());
-        userAnswerDto.setIsCorrect(userAnswer.getIsCorrect());
+//        ListenAnswerDto listenAnswerDto = listenAnswerMapper.convertModelToDto(userAnswer.getListenAnswer());
```
[the remark this answers is not in the document]
Comment (minhtrungs013) [reply src/main/java/com/tip/dg4/toeic_exam/mappers/UserAnswerMapper.java:32]: I fix it

**Pull request merged**